source/How-To-Guides/Launch-file-different-formats.rst

```diff
@@ -10,12 +10,321 @@ Using Python, XML, and YAML for ROS 2 Launch Files
+Examples
+--------
+
+Taking arguments
+^^^^^^^^^^^^^^^^
+
+Each launch file performs the following actions:
+
+* Declares arguments with and without defaults.
+* Logs a message that interpolates those arguments.
+
+.. tabs::
+
+   .. group-tab:: Python
+
+      .. code-block:: python
+
+        # example_launch.py
+
+        from launch import LaunchDescription
+        from launch.actions import DeclareLaunchArgument
+        from launch.actions import LogInfo
+        from launch.substitutions import LaunchConfiguration
+
+        def generate_launch_description():
+            return LaunchDescription([
+                DeclareLaunchArgument('who'),
+                DeclareLaunchArgument('where', default_value='home'),
+
+                LogInfo(msg=[LaunchConfiguration('who'), ' is at ', LaunchConfiguration('where')])
+            ])
+
+
+   .. group-tab:: XML
+
+      .. code-block:: xml
+
+        <!-- example_launch.xml -->
+
+        <launch>
+            <arg name="who"/>
+            <arg name="where" default="home"/>
+
+            <log message="$(var who) is at $(var where)"/>
+        </launch>
+
+   .. group-tab:: YAML
+
+      .. code-block:: yaml
+
+        # example_launch.yaml
+
+        launch:
+        - arg:
+            name: "who"
+        - arg:
+            name: "where"
+            default: "home"
+        - log:
+            message: "$(var who) is at $(var where)"
+
+Requiring processes
+^^^^^^^^^^^^^^^^^^^
+
+Each launch file performs the following actions:
+
+* Declares an argument with defaults.
+* Sleeps for some time, then shuts down launch.
+
+.. tabs::
+
+   .. group-tab:: Python
+
+      .. code-block:: python
+
+        # example_launch.py
+
+        from launch import LaunchDescription
+        from launch.actions import DeclareLaunchArgument
+        from launch.actions import ExecuteProcess
+        from launch.actions import LogInfo
+        from launch.actions import Shutdown
+        from launch.substitutions import LaunchConfiguration
+
+        def generate_launch_description():
+            return LaunchDescription([
+                DeclareLaunchArgument('duration', default_value='10'),
+                ExecuteProcess(
+                    cmd=['sleep', LaunchConfiguration('duration')],
+                    on_exit=[
+                        LogInfo(msg='Done sleeping!'),
+                        Shutdown(reason='main process terminated')
+                    ]
+                ),
+                LogInfo(msg=['Sleeping for ', LaunchConfiguration('duration'), ' seconds']),
+            ])
+
+   .. group-tab:: XML
+
+      .. code-block:: xml
+
+        <!-- Currently unsupported -->
+
+   .. group-tab:: YAML
+
+      .. code-block:: yaml
+
+        # Currently unsupported
+
+Replicating hierarchies
+^^^^^^^^^^^^^^^^^^^^^^^
+
+Each launch file performs the following actions:
+
+* Declares an argument without defaults.
+* Generates namespaced groups iteratively based on that argument.
```

Comment:
I don't really understand what this means.

Also one example that would be really helpful somewhere would be to show how to have an optional namespace.

Reply:
@jbarry-bdai What is an optional namespace?